runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkProcessContext.java

```diff
@@ -125,8 +164,50 @@ public InputT element() {
 
   @Override
   public void outputWithTimestamp(OutputT output, Instant timestamp) {
-    output(WindowedValue.of(output, timestamp,
-        windowedValue.getWindows(), windowedValue.getPane()));
+    if (windowedValue == null) {
+      // this is start/finishBundle.
+      output(noElementWindowedValue(output, timestamp, windowingStrategy));
+    } else {
+      output(WindowedValue.of(output, timestamp, windowedValue.getWindows(),
+          windowedValue.getPane()));
+    }
+  }
+
+  static <T> WindowedValue<T> noElementWindowedValue(final T output,
+                                                     final Instant timestamp,
+                                                     WindowingStrategy<?, ?> ws) {
+    @SuppressWarnings("unchecked")
+    WindowFn<Object, ?> windowFn = (WindowFn<Object, ?>) ws.getWindowFn();
```

Comment:
From my reading, the only piece of `WindowingStrategy` used is the `WindowFn`. On the one hand, passing the whole `WindowingStrategy` makes it easy to drop new fields onto the parameter with backwards compatibility, but on the other hand, it means you have greater potential dependencies on the other fields.